server/market/market.go

```diff
@@ -361,109 +472,212 @@ func (m *Market) Book() (epoch int64, buys, sells []*order.LimitOrder) {
+			m.orderFeedMtx.RLock()
+			feeds := m.orderFeeds
+			m.orderFeedMtx.RUnlock()
+			for _, s := range feeds {
+				s <- sig
+			}
+		}
+	}()
 
 	// Start the closed epoch pump, which drives preimage collection and orderly
 	// epoch processing.
 	eq := newEpochPump()
-	var wg sync.WaitGroup
-	wg.Add(1)
+	wgEpochs.Add(1)
 	go func() {
-		defer wg.Done()
+		defer wgEpochs.Done()
 		eq.Run(ctxRun)
 	}()
 
 	// Start the closed epoch processing pipeline.
-	wg.Add(1)
+	wgEpochs.Add(1)
 	go func() {
-		defer wg.Done()
-		for {
-			select {
-			case ep, ok := <-eq.ready:
-				if !ok {
-					return
-				}
-
-				// epochStart has completed preimage collection.
-				m.processReadyEpoch(ctxRun, ep)
-			case <-ctxRun.Done():
-				return
-			}
+		defer wgEpochs.Done()
+		for ep := range eq.ready {
+			// epochStart has completed preimage collection.
+			m.processReadyEpoch(ep, notifyChan)
 		}
+		log.Debugf("epoch pump drained for market %s", m.marketInfo.Name)
+		// There must be no more notify calls.
 	}()
 
-	nextEpochIdx := atomic.LoadInt64(&m.startEpochIdx)
+	m.epochMtx.Lock()
+	nextEpochIdx := m.startEpochIdx
 	if nextEpochIdx == 0 {
 		log.Warnf("Run: startEpochIdx not set. Starting at the next epoch.")
 		now := encode.UnixMilli(time.Now())
 		nextEpochIdx = 1 + now/int64(m.EpochDuration())
+		m.startEpochIdx = nextEpochIdx
 	}
+	m.epochMtx.Unlock()
+
 	epochDuration := int64(m.marketInfo.EpochDuration)
 	nextEpoch := NewEpoch(nextEpochIdx, epochDuration)
 	epochCycle := time.After(time.Until(nextEpoch.Start))
 
-	var running bool
 	var currentEpoch *EpochQueue
 	cycleEpoch := func() {
 		if currentEpoch != nil {
 			// Process the epoch asynchronously since there is a delay while the
 			// preimages are requested and clients respond with their preimages.
-			m.enqueueEpoch(eq, currentEpoch)
+			if !m.enqueueEpoch(eq, currentEpoch) {
+				return
+			}
 
 			// The epoch is closed, long live the epoch.
-			sig := &bookUpdateSignal{
-				action:   newEpochAction,
-				epochIdx: nextEpoch.Epoch,
+			sig := &updateSignal{
+				action: newEpochAction,
+				data:   sigDataNewEpoch{idx: nextEpoch.Epoch},
 			}
-			m.notify(ctxRun, sig)
+			notifyChan <- sig
 		}
-		currentEpoch = nextEpoch
 
-		nextEpoch = NewEpoch(currentEpoch.Epoch+1, epochDuration)
-		epochCycle = time.After(time.Until(nextEpoch.Start))
+		// Guard activeEpochIdx and suspendEpochIdx.
+		m.epochMtx.Lock()
+		defer m.epochMtx.Unlock()
+
+		// Check suspendEpochIdx and suspend if the just-closed epoch idx is the
+		// suspend epoch.
+		if m.suspendEpochIdx == nextEpoch.Epoch-1 {
+			epochCloseTime := encode.UnixMilli(currentEpoch.End)
+
+			// Reject incoming orders.
+			currentEpoch = nil
+			m.activeEpochIdx = 0
+
+			// Signal to the book router of the suspend:
+			notifyChan <- &updateSignal{
+				action: suspendAction,
+				data: sigDataSuspend{
+					finalEpoch:  m.suspendEpochIdx,
+					stopTime:    epochCloseTime,
+					persistBook: m.persistBook,
+				},
+			}
```

Comment:
This may not be necessary at all.  It all depends on if the book router should stay running when the feeding market suspends, and how resume is implemented.  My current thinking is that the DEX manager would handle the resume process, restarting the subsustems as it sees fit.